SteamKit2/SteamKit2/Steam/CDN/LancacheDetector.cs

```diff
@@ -0,0 +1,86 @@
+﻿/*
+ * This file is subject to the terms and conditions defined in
+ * file 'license.txt', which is part of this source code package.
+ */
+
+using System;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Sockets;
+using System.Threading.Tasks;
+
+namespace SteamKit2.Steam.CDN
+{
+    /// <summary>
+    /// Attempts to automatically resolve the Lancache's IP address in the same manner that the Steam client does.
+    ///
+    /// Will automatically try to detect the Lancache through the poisoned DNS entries.
+    /// This is a modified version from the original source : https://github.com/tpill90/lancache-prefill-common/blob/main/dotnet/LancacheIpResolver.cs
+    /// </summary>
+    public static class LancacheDetector
+    {
+        private static string TriggerDomain = "lancache.steamcontent.com";
+
+        public static async Task<bool> DetectLancacheServerAsync(HttpClient httpClient)
+        {
+            // Gets a list of ipv4 addresses, Lancache cannot use ipv6 currently
+            var ipAddresses = (await Dns.GetHostAddressesAsync( TriggerDomain ) )
+                .Where(e => e.AddressFamily == AddressFamily.InterNetwork)
+                .ToArray();
+
+            // If there are no private IPs, then there can't be a Lancache instance.  Lancache's IP must resolve to an RFC 1918 address
+            if (!ipAddresses.Any(e => e.IsPrivateAddress()))
+            {
+                return false;
+            }
+
+            // DNS hostnames can possibly resolve to more than one IP address (one-to-many), so we must check each one for a Lancache server
+            foreach (var ip in ipAddresses)
+            {
+                try
+                {
+                    // If the IP resolves to a private subnet, then we want to query the Lancache server to see if it is actually there.
+                    // Requests that are served from the cache will have an additional header.
+                    var response = await httpClient.GetAsync(new Uri($"http://{ip}/lancache-heartbeat"));
```

Comment:
This seems like more extra logic that Steam wouldn't do.

Reply:
Yeah, this is not something that Steam does.  This avoids the config error of where there is no actual cache server pointing at the IP that `lancache.steamcontent.com` resolves to, which causes Steam to fail on every request.

If we're shooting for 1:1 behavior match with Steam then I'll drop this.  Would simplify other things too.

Reply:
Yes I don't think it's our problem. And it can be solved by making it optional (by having a call to enable lancache).

Reply:
So drop the whole class or just that piece of functionality